sqlglot/expressions.py

```diff
@@ -766,6 +766,10 @@ class Hint(Expression):
     arg_types = {"expressions": True}
 
 
+class TableHint(Expression):
```

Comment:
Spark calls this "join hints".

And I imagine other dialects have totally different types of hints.

I'm worried we're making the core expression too Spark specific. Can we parse these expressions into `Hint`, but have the Spark dialect handle parsing the expressions as `Table`s?

Reply:
Good to rename to `JoinHint` if we continue with this approach. 👍 

So `Hint` doesn't actually represent a single hint... it represents a collection of hints. Those individual hints prior to this PR were Anonymous functions and then this PR allows us to differentiate the anonymous function ones (which are handled fine since they contain columns) and the join hint ones which behave differently since they contain tables. 

In my mind if we know at parsing time that the hint being expressed is one of these two types then we should put it in an expression that represents that. I did though stop short of implementing though that completely since the anonymous hints aren't causing any issues currently.

Reply:
Ah, got it. 

Is it possible to have a hint with tables and columns? The Spark ones don't, but maybe some other dialect does? Or maybe Spark will _eventually_?

Reply:
Maybe but just some googling suggests we don't have another example to compare against. Generally my preference here is to model the solution for the only given example we have and then iterate as new patterns emerge.

Reply:
agreed

Reply:
`coalesce` is a PartitionHint which this PR leaves as an Anonymous function since currently they are handled fine in the codebase as one.

Reply:
I'm going to stick with the current approach but renamed everything to `JoinHint`. To me this is a good first step and we can generalize as we get more examples (if we ever do).